### DISCUSSION TITLE
Why does nested_chats only inherit the last_msg of historical information and discord messages when nested_chats finish the task?

### DISCUSSION THREAD

**Discussion by smallQQ0227** [microsoft/autogen#4132] [General]
ConversableAgent._get_chats_to_run()  is responsible for passing the last message to nested_chats from the registered agents, after which the agents in nested_chats begin their conversation. However, there are two issues: 
1. Other historical messages outside of nested_chats are not passed into nested_chats(Similar to the  [#issue 3314](https://github.com/microsoft/autogen/issues/3314)). This is easy to solve. We just need to modify one line of code:
```python
@staticmethod
    def _get_chats_to_run(
        chat_queue: List[Dict[str, Any]], recipient: Agent, messages: Union[str, Callable], sender: Agent, config: Any
    ) -> List[Dict[str, Any]]:
        ###### The following line of code is the original code, which just passes the last message from the registered agent.########
        # last_msg = messages[-1].get("content")
        ######Simply embed the historical information and the last message into a simple prompt template. #########
        history=json.dumps(messages[:-1])
        last_msg=json.dumps(messages[-1])
        last_msg=f"""Context:{history} . 
                    The latest reply:{last_msg} .
                    Please provide the correct and concise answer.
            """
        chat_to_run = []
        for i, c in enumerate(chat_queue):
            current_c = c.copy()
            if current_c.get("sender") is None:
                current_c["sender"] = recipient
            message = current_c.get("message")
            # If message is not provided in chat_queue, we by default use the last message from the original chat history as the first message in this nested chat (for the first chat in the chat queue).
            # NOTE: This setting is prone to change.
            if message is None and i == 0:
                message = last_msg

            if callable(message):
                message = message(recipient, messages, sender, config)
            # We only run chat that has a valid message. NOTE: This is prone to change dependin on applications.
            if message:
                current_c["message"] = message
                chat_to_run.append(current_c)
        return chat_to_run
```
However, this method will print out all historical information when nested_chats starts the conversation, which is not aesthetically pleasing.
2、When the nested_chats finished the task and return the final result, the agents in the nested_chats will diacord all the history messages. This means that when nested_chats is triggered again (printing 'Starting a new chat'), we need to ensure that all important historical information is passed to nested_chats. Therefore, is there a way for the agents in nested_chats to retain memory, or to allow agents inside and outside of nested_chats to share the same message_pool?

**Comment by ekzhu** [comment]
Thanks for the feedback. We will address this issue in the v0.4 version. 

Please stay turned.

**Comment by ekzhu** [comment]
Latest update from v0.4 preview: the SocietyofMindAgent is a built-in approach to nested chat. 

https://microsoft.github.io/autogen/dev/reference/python/autogen_agentchat.agents.html#autogen_agentchat.agents.SocietyOfMindAgent

You can refer to its implementation and implement your own nested chat agent using custom agent: 

https://microsoft.github.io/autogen/dev/user-guide/agentchat-user-guide/tutorial/custom-agents.html

Will be released in the upcoming dev12 release.
